pkg/operators/hub/controller.go

```diff
@@ -2,34 +2,295 @@ package hub
-	"k8s.io/client-go/informers"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
 
+	"github.com/openshift/api"
+	"github.com/openshift/library-go/pkg/assets"
 	"github.com/openshift/library-go/pkg/controller/factory"
 	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	nucleusv1client "github.com/open-cluster-management/api/client/nucleus/clientset/versioned/typed/nucleus/v1"
+	nucleusinformer "github.com/open-cluster-management/api/client/nucleus/informers/externalversions/nucleus/v1"
+	nucleuslister "github.com/open-cluster-management/api/client/nucleus/listers/nucleus/v1"
+	nucleusapiv1 "github.com/open-cluster-management/api/nucleus/v1"
+	"github.com/open-cluster-management/nucleus/pkg/helpers"
+	"github.com/open-cluster-management/nucleus/pkg/operators/hub/bindata"
+)
+
+var (
+	genericScheme = runtime.NewScheme()
+	genericCodecs = serializer.NewCodecFactory(genericScheme)
+	genericCodec  = genericCodecs.UniversalDeserializer()
+	crdNames      = []string{
+		"manifestworks.work.open-cluster-management.io",
+		"spokeclusters.cluster.open-cluster-management.io",
+	}
 )
 
+const (
+	nucleusHubFinalizer     = "nucleus.open-cluster-management.io/hub-core-cleanup"
+	nucluesHubCoreNamespace = "open-cluster-management"
+	hubCoreApplied          = "Applied"
+	hubCoreAvailable        = "Available"
+)
+
+func init() {
+	utilruntime.Must(api.InstallKube(genericScheme))
+}
+
 type nucleusHubController struct {
-	kubeClient         kubernetes.Interface
-	apiExtensionClient apiextensionsclient.Interface
+	nucleusClient                  nucleusv1client.HubCoreInterface
+	nucleusLister                  nucleuslister.HubCoreLister
+	kubeClient                     kubernetes.Interface
+	apiExtensionClient             apiextensionsclient.Interface
+	currentHubDeploymentGeneration int64
 }
 
 // NewNucleusHubController construct nucleus hub controller
 func NewNucleusHubController(
 	kubeClient kubernetes.Interface,
 	apiExtensionClient apiextensionsclient.Interface,
-	kubeInformers informers.SharedInformerFactory,
+	nucleusClient nucleusv1client.HubCoreInterface,
+	nucleusInformer nucleusinformer.HubCoreInformer,
 	recorder events.Recorder) factory.Controller {
 	controller := &nucleusHubController{
-		kubeClient:         kubeClient,
-		apiExtensionClient: apiExtensionClient,
+		kubeClient:                     kubeClient,
+		apiExtensionClient:             apiExtensionClient,
+		nucleusClient:                  nucleusClient,
+		nucleusLister:                  nucleusInformer.Lister(),
+		currentHubDeploymentGeneration: 0,
```

Comment:
this will cause excessive rollouts. I suspect you want the OperatorStatus from openshift/api to track it.